site/docs/api/error-handling.md

```diff
@@ -0,0 +1,136 @@
+
+In Jet we also have generic options for dealing with random errors
+affecting our jobs.
+
+### Jobs Without Mutable State
+
+For many streaming jobs, specifically the ones which don't have any
+processing guarantee configured, the pipeline definition and the job
+config are the only parts we can identify as state, and those are
+immutable.
+
+Batch jobs also fall into the category of immutable state jobs.
+
+One option for dealing with failure in immutable state jobs is simply
+restarting them (once the cause of the failure has been addressed).
+Restarted streaming jobs lacking mutable state can just resume
+processing the input data flow from the current point in time, batch
+jobs can be re-run from the beginning.
+
+### Processing Guarantees
+
+Jobs with mutable state, the ones that have [processing
+guarantees](../architecture/fault-tolerance#processing-guarantee-is-a-shared-concern)
+set, achieve their fault tolerance by periodically saving [recovery
+snapshots](../architecture/fault-tolerance#distributed-snapshot). When
+such jobs fail, not only does their execution stop, but also their
+snapshots get deleted. This makes it impossible to resume them without
+loss.
+
+To cope with this situation, starting from Jet 4.3, jobs can be
+configured to be only suspended when failures occur, instead of having
+them fail completely. For details see
+[`JobConfig.setSuspendOnFailure`](/javadoc/{jet-version}/com/hazelcast/jet/config/JobConfig.html#setSuspendOnFailure(boolean)).
+Note: to preserve behavior from older versions, this configuration
+option is not the default, so it must be enabled explicitly.
+
+Having them in a suspended state preserves their snapshots and makes
+resumptions possible, once the root cause of the failure has been
+addressed.
+
+In the open-source version of Jet this scenario is limited to
+fixing the input data by some external means and then simply [resuming
+the job](../operations/job-management#restarting) via the client API.
+
+Enterprise users of Jet have the added option of [job upgrades](../enterprise/job-update).
+They can:
+
+* export the latest snapshot
+* update their pipeline, if needed, for example, to cope with unexpected
+  data
+* resubmit a new job based on the exported snapshot and the updated pipeline
+
+One caveat of the suspend-on-failure feature is that the latest snapshot
+is not a "failure snapshot." Jet can't take a full snapshot right at the
+moment of the failure, because its processors can produce accurate
+snapshots only when in a healthy state. So the latest snapshot we will
+have available will be whatever has been saved at the last periodic
+snapshotting phase. But even so, the recovery procedure should preserve
+at-least-once guarantees.
+
+More information on the topic can also be found in the feature's design
+document, for example, how clients that have submitted the job can
+[discover if it got suspended this
+way](../design-docs/012-improved-job-resilience#notifying-the-client).
```

Comment:
We should not refer to description of important features in TDD. We can refer to TDD to read more why the design is as it is or so.